src/Datadog.Trace/Configuration/JsonConfigurationSource.cs

```diff
@@ -97,5 +99,40 @@ public T GetValue<T>(string key)
                        ? default
                        : token.Value<T>();
         }
+
+        /// <summary>
+        /// Gets a <see cref="ConcurrentDictionary{TKey, TValue}"/> containing all of the values.
+        /// </summary>
+        /// <remarks>
+        /// Example JSON where `globalTags` is the configuration key.
+        /// {
+        ///  "globalTags": {
+        ///     "name1": "value1",
+        ///     "name2": "value2"
+        ///     }
+        /// }
+        /// </remarks>
+        /// <param name="key">The key that identifies the setting.</param>
+        /// <returns><see cref="IDictionary{TKey, TValue}"/> containing all of the key-value pairs.</returns>
+        /// <exception cref="JsonReaderException">Thrown if the configuration value is not a valid JSON string.</exception>"
+        public IDictionary<string, string> GetDictionary(string key)
+        {
+            var token = _configuration.SelectToken(key, errorWhenNoMatch: false);
+            if (token == null)
+            {
+                return null;
+            }
+
+            // Presence of a curly brace indicates that this is likely a JSON string. An exception will be thrown
+            // if it fails to parse or convert to a dictionary.
+            if (token.ToString().Contains("{"))
```

Comment:
At this point the json library already deserialized the json string, so I don't think we should serialize it back to a json string by calling `JToken.ToString()`. Consider checking the `JToken.Type` instead.
```suggestion
            if (token.Type == JTokenType.Object)
```

Reply:
oh that's interesting. I was looking for a `JToken` method or property to validate, but that looks good! Thanks I'll test it out.

Reply:
@lucaspimentel `if (token.Type == JTokenType.Object)` won't work. The `token.Type` is `String` even in the case where the config string contains valid JSON, so it is never converted. What might work best is to not convert it using a `ToObject` but instead always treat it as a string and use `StringConfigurationSource.ParseCustomKeyValues` all of the time. This has the side-benefit of being able to handle the case of an extra curly brace in the config.

Reply:
Ok, this has finally been improved (thanks for the insight, @lucaspimentel ). Now the tests for `JsonConfigurationSource` do not start by putting the key/value data into a dictionary. Instead, the value is passed directly to the `JsonConfigurationSource` constructor, as it should.

I also added a second bad data test.